### PR TITLE
[rllib] Initial RLLib documentation

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: doc/source/rllib.rst [hunk 1]
```diff
@@ -0,0 +1,112 @@
+RLLib: Ray's modular and scalable reinforcement learning library
```
Comment (ericl) [inline doc/source/rllib.rst]: I changed my mind about modular, how about just scalable :)

File: doc/source/rllib.rst [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+Examples
+--------
+
+Some good hyperparameters and settings are available in
+`the repository <https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/tuned_examples.sh>`__
+(some of them are tuned to run on GPUs). If you find better settings or tune
+an algorithm on a different domain, consider submitting a Pull Request!
+
+The API
```
Comment (ericl) [inline doc/source/rllib.rst]: I would separate this into two sections: the public API for those that just want to run different RL algorithms, and the internal APIs for those that want to contribute to RLllib.

File: doc/source/rllib.rst [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+Getting Started
+---------------
+
+You can run training with
+
+::
+
+    python ray/python/ray/rllib/train.py --env CartPole-v0 --alg PPO
+
+
```
Comment (ericl) [inline doc/source/rllib.rst]: Btw could we reduce the default PPO batch size? It OOMs my laptop due to the memory requirements, so perhaps its not a very friendly example.

File: doc/source/rllib.rst [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+and registered by the user.
+
+Getting Started
+---------------
+
+You can run training with
+
+::
+
+    python ray/python/ray/rllib/train.py --env CartPole-v0 --alg PPO
```
Comment (ericl) [inline doc/source/rllib.rst]: Perhaps also mention that results are logged to /tmp/ray, and that agents print out their log directory?

There are also some relevant files:
- result.json
- config.json
- tensorboard events (view with tensorboard --logdir=/tmp/ray)

File: doc/source/rllib.rst [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+   borrows code from
+   `here <https://github.com/openai/evolution-strategies-starter>`__.
+
+-  `Deep Q Networks <https://www.cs.toronto.edu/~vmnih/docs/dqn.pdf>`__
+   based on `OpenAI baselines <https://github.com/openai/baselines>`__.
+
+-  `The Asynchronous Advantage Actor-Critic <https://arxiv.org/abs/1602.01783>`__
+   based on `the OpenAI starter agent <https://github.com/openai/universe-starter-agent>`__.
+
+These algorithms can be run on any OpenAI gym MDP, including custom ones written
```
Comment (ericl) [inline doc/source/rllib.rst]: Should we say something about their scalability here? E.g. ES scales to clusters with thousands of cores, A3C to dozens of cores, PPO to hundreds of cores + several GPUs.

File: doc/source/rllib.rst (hunk 1 above)
Comment (robertnishihara) [reply doc/source/rllib.rst]: Sounds good to me

File: doc/source/conf.py [hunk 6]
```diff
@@ -18,8 +18,14 @@
...
 import mock
-MOCK_MODULES = ["pyarrow",
+MOCK_MODULES = ["gym",
+                "tensorflow",
+                "tensorflow.contrib",
+                "tensorflow.contrib.slim",
+                "tensorflow.contrib.rnn",
+                "pyarrow",
                 "pyarrow.plasma",
+                "smart_open",
```
Comment (robertnishihara) [inline doc/source/conf.py]: This can be a separate PR, but we should probably modify `setup.py` so people can do `pip install ray[rllib]` to get the relevant dependencies. We can also do `pip install ray[webui]` (see #970).

File: doc/source/rllib.rst [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+-  `Proximal Policy Optimization <https://arxiv.org/abs/1707.06347>`__ which
+   is a proximal variant of `TRPO <https://arxiv.org/abs/1502.05477>`__.
+
+-  Evolution Strategies which is decribed in `this
+   paper <https://arxiv.org/abs/1703.03864>`__. Our implementation
+   borrows code from
+   `here <https://github.com/openai/evolution-strategies-starter>`__.
+
+-  `Deep Q Networks <https://www.cs.toronto.edu/~vmnih/docs/dqn.pdf>`__
+   based on `OpenAI baselines <https://github.com/openai/baselines>`__.
```
Comment (robertnishihara) [inline doc/source/rllib.rst]: Should we add this in later when it is more polished?

File: doc/source/rllib.rst [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+Some good hyperparameters and settings are available in
+`the repository <https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/tuned_examples.sh>`__
+(some of them are tuned to run on GPUs). If you find better settings or tune
+an algorithm on a different domain, consider submitting a Pull Request!
+
+The API
+-------
+
+Note that the API is not considered to be stable yet.
+
```
Comment (robertnishihara) [inline doc/source/rllib.rst]: I think people would prefer to use the Python API (instead of calling `train.py`), so we should document that or at least tell people to look at `train.py` to figure out how to use the Python API.

File: doc/source/rllib.rst (hunk 1 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Yeah, I also don't think we should claim modular just yet ;)

Comment (robertnishihara) [reply doc/source/rllib.rst]: Oh, also we need to document how to use it in the cluster setting :)

File: doc/source/rllib.rst (hunk 5 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Done

File: doc/source/rllib.rst (hunk 4 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Done

File: doc/source/rllib.rst (hunk 3 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: I reduced it to 10000, does it work on your laptop now?

File: doc/source/rllib.rst (hunk 2 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Done

File: doc/source/rllib.rst (hunk 7 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Done

File: doc/source/rllib.rst (hunk 8 above)
Comment (pcmoritz) [reply doc/source/rllib.rst]: Done

**Comment by robertnishihara** [comment]
Looks good to me.
